File: ts-sdk/src/providers/openai/adapter.ts
import { CallContext } from '@/types/context.js';
import { LLMRequest } from '@/types/request.js';
import { LLMResponse } from '@/types/response.js';
import { messagesToOpenAIInput, contentFromOpenAI, usageFromOpenAI } from './common.js';

interface OpenAIResponsesClient {
  responses: {
    create: (args: Record<string, unknown>) => Promise<unknown>;
  };
}

interface OpenAIChatClient {
  chat: {
    completions: {
      create: (args: Record<string, unknown>) => Promise<unknown>;
    };
  };
}

function isOpenAIResponsesClient(client: unknown): client is OpenAIResponsesClient {
  if (!client || typeof client !== 'object' || !('responses' in client)) return false;
  const responses = (client as { responses?: { create?: unknown } }).responses;
  return typeof responses?.create === 'function';
}

function isOpenAIChatClient(client: unknown): client is OpenAIChatClient {
  if (!client || typeof client !== 'object' || !('chat' in client)) return false;
  const chat = (client as { chat?: { completions?: { create?: unknown } } }).chat;
  return typeof chat?.completions?.create === 'function';
}

function responsesArgs(request: LLMRequest, defaultModel?: string): Record<string, unknown> {
  const model = request.model ?? defaultModel;
  if (!model) throw new Error('No model provided (set request.model or defaultModel).');
  const params = { ...(request.params ?? {}) };
  delete params.model;
  delete params.input;
  return { model, input: messagesToOpenAIInput(request), ...params };
}

function chatArgs(request: LLMRequest, defaultModel?: string): Record<string, unknown> {
  const model = request.model ?? defaultModel;
  if (!model) throw new Error('No model provided (set request.model or defaultModel).');
  const params = { ...(request.params ?? {}) };
  delete params.model;
  delete params.messages;
  return { model, messages: messagesToOpenAIInput(request), ...params };
}

export class OpenAIResponsesAdapter {
  private readonly client: OpenAIResponsesClient;
  private readonly defaultModel?: string;

  constructor(client: unknown, opts?: { defaultModel?: string }) {
    if (!isOpenAIResponsesClient(client)) {
      throw new Error('Client does not have responses API');
    }
    this.client = client;
    this.defaultModel = opts?.defaultModel;
  }

  async call(request: LLMRequest, _ctx: CallContext): Promise<LLMResponse> {
    const raw = await this.client.responses.create(responsesArgs(request, this.defaultModel));
    return { content: contentFromOpenAI(raw), usage: usageFromOpenAI(raw), raw };
  }
}

export class OpenAIChatCompletionsAdapter {
  private readonly client: OpenAIChatClient;
  private readonly defaultModel?: string;

  constructor(client: unknown, opts?: { defaultModel?: string }) {
    if (!isOpenAIChatClient(client)) {
      throw new Error('Client does not have chat.completions API');
    }
    this.client = client;
    this.defaultModel = opts?.defaultModel;
  }

  async call(request: LLMRequest, _ctx: CallContext): Promise<LLMResponse> {
    const raw = await this.client.chat.completions.create(chatArgs(request, this.defaultModel));
    return { content: contentFromOpenAI(raw), usage: usageFromOpenAI(raw), raw };
  }
}

File: python-sdk/axon/providers/openai/adapter.py
from __future__ import annotations

from typing import Any

from axon.providers.openai._common import (
    content_from_openai,
    messages_to_openai_input,
    usage_from_openai,
)
from axon.types import LLMRequest, LLMResponse


def _call_kwargs(request: LLMRequest, default_model: str | None) -> dict[str, Any]:
    model = request.model or default_model
    if not model:
        raise ValueError("No model provided (set request.model or default_model).")

    params = dict(request.params)
    params.pop("model", None)
    params.pop("input", None)

    return {
        "model": model,
        "input": messages_to_openai_input(request),
        **params,
    }


class OpenAIResponsesAdapter:
    def __init__(self, client: Any, *, default_model: str | None = None) -> None:
        self._client = client
        self._default_model = default_model

    def call(self, request: LLMRequest, ctx: Any) -> LLMResponse:
        resp = self._client.responses.create(
            **_call_kwargs(request, self._default_model)
        )
        return LLMResponse(
            content=content_from_openai(resp),
            usage=usage_from_openai(resp),
            raw=resp,
        )


class AsyncOpenAIResponsesAdapter:
    def __init__(self, client: Any, *, default_model: str | None = None) -> None:
        self._client = client
        self._default_model = default_model

    async def acall(self, request: LLMRequest, ctx: Any) -> LLMResponse:
        resp = await self._client.responses.create(
            **_call_kwargs(request, self._default_model)
        )
        return LLMResponse(
            content=content_from_openai(resp),
            usage=usage_from_openai(resp),
            raw=resp,
        )


def _chat_kwargs(request: LLMRequest, default_model: str | None) -> dict[str, Any]:
    model = request.model or default_model
    if not model:
        raise ValueError("No model provided (set request.model or default_model).")

    params = dict(request.params)
    params.pop("model", None)
    params.pop("messages", None)

    return {
        "model": model,
        "messages": messages_to_openai_input(request),
        **params,
    }


class OpenAIChatCompletionsAdapter:
    def __init__(self, client: Any, *, default_model: str | None = None) -> None:
        self._client = client
        self._default_model = default_model

    def call(self, request: LLMRequest, ctx: Any) -> LLMResponse:
        resp = self._client.chat.completions.create(
            **_chat_kwargs(request, self._default_model)
        )
        return LLMResponse(
            content=content_from_openai(resp),
            usage=usage_from_openai(resp),
            raw=resp,
        )


class AsyncOpenAIChatCompletionsAdapter:
    def __init__(self, client: Any, *, default_model: str | None = None) -> None:
        self._client = client
        self._default_model = default_model

    async def acall(self, request: LLMRequest, ctx: Any) -> LLMResponse:
        resp = await self._client.chat.completions.create(
            **_chat_kwargs(request, self._default_model)
        )
        return LLMResponse(
            content=content_from_openai(resp),
            usage=usage_from_openai(resp),
            raw=resp,
        )
